File: 2147-number-of-ways-to-divide-a-long-corridor/2147-number-of-ways-to-divide-a-long-corridor.py
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        N = len(corridor)
        MOD = 10 ** 9 + 7
        ans = 1
        
        count = 0
        left = corridor.count("S")
        if left % 2 != 0 or N < 2 or left == 0:
            return 0
        
        streak = 0
        for i in range(N):
            if corridor[i] == "S":
                count += 1
                left -= 1
            
            if count == 2:
                streak += 1
            
            elif count < 2:
                if streak > 0:
                    ans *= streak
                streak = 0
            else:
                if streak > 0:
                    ans *= streak
                count = 1
                streak = 0
            ans %= MOD
        
        return ans
    

File: 2147-number-of-ways-to-divide-a-long-corridor/2147-number-of-ways-to-divide-a-long-corridor.cpp
#include <cmath>

class Solution {
public:
    int numberOfWays(string corridor) {
        int N = corridor.length();
        int p_count = 0;
        int MOD = pow(10, 9) + 7;
        for (int i = 0; i <= N; i++){
            if (corridor[i] == 'S'){
                p_count++;
            }
        }
        
        if (p_count % 2 != 0 || p_count < 2){
            return 0;
        }
        
        long long ans = 1;
        int streak = 0;
        int curr_count = 0;
        for (int i = 0; i <= N; i++){
            if (corridor[i] == 'S'){
                curr_count++;
            }
            
            if (curr_count == 2){
                streak++;
            }
            if (curr_count > 2){
                ans *= streak;
                ans %= MOD;
                curr_count = 1;
                streak = 0;
            }
        }
        
        return ans;
    }
};